astropy/utils/tests/test_decorators.py

```diff
@@ -769,3 +770,109 @@ class TestClass:
         pass
 
     assert inspect.getdoc(TestClass) == 'what we do is strange.'
+
+
+def make_on_metaclass_examples():
+    class ExampleMeta(type):
+        def method(cls):
+            return "default"
+
+    class ExampleNoMethod(metaclass=ExampleMeta):
+        expect = "default"
+
+    class ExampleOnMeta(metaclass=ExampleMeta):
+        expect = "ExampleOnMeta"
+
+        @on_metaclass
+        def method(cls):
+            return cls.__name__
+
+    class ExampleMetaclassMethod(metaclass=ExampleMeta):
+        expect = "ExampleMeta"
+
+        @on_metaclass
+        @classmethod
+        def method(cls):
+            return cls.__name__
+
+    class ExampleClassProperty(metaclass=ExampleMeta):
+        expect = "ExampleClassProperty"
+
+        @on_metaclass
+        @property
+        def method(cls):
+            return cls.__name__
+
+    class ExampleMetaClassProperty(metaclass=ExampleMeta):
+        expect = "ExampleMeta"
+
+        @on_metaclass
+        @classmethod
+        @property
+        def method(cls):
+            return cls.__name__
+
+    classes = [ExampleNoMethod, ExampleOnMeta, ExampleMetaclassMethod, ExampleClassProperty]
+    # py3.8- doesn't allow classmethod to stack on a property.
+    if sys.version_info >= (3, 9):
+        classes.append(ExampleMetaClassProperty)
+
+    return classes
+
+
+class Test_dir_:
+    """Test `astropy.utils.decorators._dir_`."""
+
+    def test_replace(self):
+        """Test basic replacing of ``__dir__`` with ``_dir_``."""
+        class Example:
+            __dir__ = _dir_(exclude=['_private'])
+
+            @classmethod
+            def _private(cls):
+                return
+
+        # It's in the class
+        assert "_private" in dir(Example)
+
+        # But not the instance
+        ex = Example()
+        assert "_private" not in dir(ex)
```

Comment:
Is there some reason why this should be spread out on two lines?
```suggestion
        assert "_private" not in dir(Example())
```

Reply:
☑️